core/lib/multivm/src/versions/vm_latest/tests/evm_emulator.rs

```diff
@@ -74,3 +92,229 @@ fn tracing_evm_contract_deployment() {
+        .with_random_rich_accounts(1)
+        .build();
+    let account = &mut vm.rich_accounts[0];
+    let tx = account.get_l2_tx_for_execute(
+        Execute {
+            contract_address: Some(called_address),
+            calldata: vec![],
+            value: 0.into(),
+            factory_deps: vec![],
+        },
+        None,
+    );
+
+    let (_, vm_result) = vm
+        .vm
+        .execute_transaction_with_bytecode_compression(tx, true);
+    assert!(!vm_result.result.is_failed(), "{:?}", vm_result.result);
+}
+
+fn build_vm(deploy_emulator: bool, contract_address: Address) -> VmTester<HistoryEnabled> {
+    let mock_emulator = read_bytecode(MOCK_EMULATOR_PATH);
+    let mut storage = InMemoryStorage::with_system_contracts(hash_bytecode);
+    let mut system_env = default_system_env();
+    if deploy_emulator {
+        let evm_bytecode: Vec<_> = (0..=u8::MAX).collect();
+        let evm_bytecode_hash = hash_evm_bytecode(&evm_bytecode);
+        storage.set_value(get_code_key(&contract_address), evm_bytecode_hash);
+        storage.set_value(
+            get_known_code_key(&evm_bytecode_hash),
+            H256::from_low_u64_be(1),
+        );
+
+        system_env.base_system_smart_contracts.evm_emulator = Some(SystemContractCode {
+            hash: hash_bytecode(&mock_emulator),
+            code: bytes_to_be_words(mock_emulator),
+        });
+    } else {
+        let emulator_hash = hash_bytecode(&mock_emulator);
+        storage.set_value(get_code_key(&contract_address), emulator_hash);
+        storage.set_value(get_known_code_key(&emulator_hash), H256::from_low_u64_be(1));
+        storage.store_factory_dep(emulator_hash, mock_emulator);
+        // Set `isUserSpace` in the emulator storage to `true`, so that it skips emulator-specific checks
+        storage.set_value(
+            StorageKey::new(AccountTreeId::new(contract_address), H256::zero()),
+            H256::from_low_u64_be(1),
+        );
+    }
+
+    VmTesterBuilder::new(HistoryEnabled)
+        .with_system_env(system_env)
+        .with_storage(storage)
+        .with_execution_mode(TxExecutionMode::VerifyExecute)
+        .with_random_rich_accounts(1)
+        .build()
+}
+
+/// `deploy_emulator = false` here and below tests the mock emulator as an ordinary contract (i.e., sanity-checks its logic).
+#[test_casing(2, [false, true])]
+#[test]
+fn mock_emulator_with_payment(deploy_emulator: bool) {
+    let mock_emulator_abi = load_contract(MOCK_EMULATOR_PATH);
+    let recipient_address = Address::repeat_byte(0x12);
+    let mut vm = build_vm(deploy_emulator, recipient_address);
+    let account = &mut vm.rich_accounts[0];
+
+    let test_payment_fn = mock_emulator_abi.function("testPayment").unwrap();
+
+    let mut current_balance = U256::zero();
+    for i in 1_u64..=5 {
+        let transferred_value = (1_000_000_000 * i).into();
+        current_balance += transferred_value;
+        let tx = account.get_l2_tx_for_execute(
```

Comment:
I'd propose to move this logic to the separate function

https://github.com/matter-labs/zksync-era/pull/3045/files#diff-36e5f0411039c083b35af6d6273d4dba7992a214a5e6330b45bc5a451d52a7b1R190-R205